common/ethereum/proof.go

```diff
@@ -0,0 +1,168 @@
+// This file was adapted from go-ethereum. Here's the go-ethereum license reproduced:
+
+// Copyright 2014 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package trie implements Merkle Patricia Tries.
+
+package ethereum
+
+import (
+	"bytes"
+	"errors"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+func NewProof() *Proof {
```

Comment:
Could you have used map to define the Proof to simplify the Put,Get,Has,Delete etc.? map[[]byte][]byte... Or is there a reason you need to have them indexed?

Reply:
i tried map, i think cosmos sdk is allergic to maps in protobuf though due to its non-determinism.